### ISSUE TITLE
Stream.CopyTo does not respect buffer size parameter

### ISSUE THREAD

**Issue by madelson** [dotnet/runtime#72716]
### Description

`Stream.CopyTo(Stream, int)` says that it will "use the specified buffer size", but sometimes it will use a larger buffer instead.

### Reproduction Steps

```cs
static void Main()
{
	using var source = new FixedBufferStream();
	using var dest = new MemoryStream();
	source.CopyTo(dest, bufferSize: 3); // throws ArgumentException: buffer too long: 16
}

class FixedBufferStream : Stream
{
	private readonly MemoryStream _source = new MemoryStream(new byte[1000]);

	public override bool CanRead => true;
	public override bool CanSeek => false;
	public override bool CanWrite => false;

	public override long Length => throw new NotImplementedException();
	public override long Position { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

	public override void Flush()
	{
		throw new NotImplementedException();
	}

	public override int Read(byte[] buffer, int offset, int count)
	{
		if (count > 4) { throw new ArgumentException($"buffer too long: {count}"); }
		return this._source.Read(buffer, offset, count);
	}

	public override long Seek(long offset, SeekOrigin origin)
	{
		throw new NotImplementedException();
	}

	public override void SetLength(long value)
	{
		throw new NotImplementedException();
	}

	public override void Write(byte[] buffer, int offset, int count)
	{
		throw new NotImplementedException();
	}
}
```

### Expected behavior

Perhaps this is just an error in documentation, but I think the passed-in buffer size should be respected. This isn't actively causing a bug for me, it just caused some confusion/surprise when it came up during some testing.

### Actual behavior

In actuality, it uses whatever buffer size comes back from `ArrayPool<byte>.Shared.Rent(bufferSize)`, which can be longer.

### Regression?

Not that I know of.

### Known Workarounds

You can override `CopyTo`/`CopyToAsync` to provide your own implementation that respects the buffer size.

### Configuration

Saw this on .NET 5, Windows 10, although the code in question still appears to be present on main: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L70

### Other information

Seems like a trivial fix would be to just pass `bufferSize` instead of `buffer.Length` when calling `Read` (with a similar change for `CopyToAsync`).

**Comment by huoyaoyuan** [comment]
`CopyTo` asks the buffer from array pool, which round up the buffer size to power of 2.
It then uses the full buffer, instead of specified size.
There is a difference of behaviour comparing to .NET Framework. But in practice the specified buffer size is mostly power of 2.

**Comment by madelson** [comment]
@huoyaoyuan yes I agree this is the cause. The question in my mind is whether that is what it *should* do when the user passes a specific buffer size.

**Comment by stephentoub** [comment]
This should be treated as a documentation issue; it'd be fine to clarify the docs that the specified buffer size is a hint or minimum.  Artificially constraining the implementation to use only a subset of the available buffer (e.g. changing the stream.Read call to pass in the requested buffer size rather than buffer.Length) would server little purpose other than to negatively impact performance.
